pkg/util/hardware/hw_utils.go

```diff
@@ -59,6 +67,63 @@ func ParseCPUSetLine(cpusetLine string) (cpusList []int, err error) {
 	return
 }
 
+func ParseCPUSetFile(filepath string) ([]int, error) {
+	cpuset, err := util.ScanLine(filepath)
+	if err != nil {
+		return nil, err
+	}
+	cpusList, err := ParseCPUSetLine(cpuset)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cpuset file: %v", err)
+	}
+	return cpusList, nil
+}
+
+func GetPodCPUSet() ([]int, error) {
+	return ParseCPUSetFile(CPUSET_PATH)
+}
+
+func getCPUThreads(cpu int) (threads map[int]bool, err error) {
+	filepath := fmt.Sprintf(coreCPUsListFmt, cpu)
+	if _, err := os.Stat(filepath); os.IsNotExist(err) {
+		filepath = fmt.Sprintf(threadSiblingsListFmt, cpu)
+	}
+	cpuList, err := ParseCPUSetFile(filepath)
+	if err != nil {
+		return nil, err
+	}
+	threads = make(map[int]bool)
+	for _, cpu := range cpuList {
+		threads[cpu] = true
+	}
+	return
+}
+
+func GroupCPUThreads(cpus []int) (sorted []int, err error) {
+	handled := make(map[int]bool)
+	for i := 0; i < len(cpus); i++ {
+		if handled[cpus[i]] {
+			continue
+		}
+
+		sorted = append(sorted, cpus[i])
+		handled[cpus[i]] = true
+
+		threads, err := getCPUThreads(cpus[i])
+		if err != nil {
+			return nil, err
+		}
+
+		for j := i + 1; j < len(cpus); j++ {
```

Comment:
is there a way to do this that avoids a n^2 operation?

I know this is rare today, but thinking a head a bit, if someone wanted a cpu with something like 112 vcpus, there are technically single processors that can supply something like that (like  xeon platinum 9282 for example).  That's potentially ~12000 comparisons.

Reply:
I agree with David. We should consider a different approach. 
For example, you could generate a map of CPU_ID siblings ahead of time, you could create a new cpuset based on this in one loop
In case HT is disabled, this map would be just empty.

Perhaps, something along the lines of the following example:  
For 4 vCPUs  + emulator thread (pin 5 threads in total)
CPUSET could be 0,1,2,13,15
and hyper threads mapping, 0,2; 1,13; 4,15

```
	newCPUSet := []int{}
	cpuset := map[int]bool{0:true,1:true,2:true,13:true,15:true}
	siblingCPUs := map[int]int{0:2,2:0,1:13,13:1,4:15,15:4}
	for cpu := range cpuset {
		// add the item to the cpuset
		newCPUSet = append(newCPUSet, cpu)
		delete(cpuset, cpu)
		
		if sibling, exist := siblingCPUs[cpu]; exist {
			if cpuset[sibling] {
				newCPUSet = append(newCPUSet, sibling)
				delete(cpuset, sibling)	
			}
		}
	}

```

Reply:
Sure, I have no objections. I just wanted to get some feedback on the PR before doing any kind of optimizations. I will consider the comments. Thanks for the hints :+1: